### PR TITLE
[MIR] Enhance the SimplifyCfg pass to merge consecutive blocks

### PR DESCRIPTION
**Pull request by gereeter**
This is a rebase of @dotdash's #29838. I'm not sure that I like the last commit, as it seems a lot harder to tell that predecessors are tracked correctly when what predecessors are changing isn't written down.

### CODE REVIEW

**Comment by dotdash** [comment]
Thanks for picking that up! I didn't have the time to take care of it yet. I'm not sure if I like the predecessor counter instead of the map though. I'm somewhat expecting people to come up with passes that will require actual predecessor lookup, and then it'd be nice if the existing code already deals with keeping the map updated.

Did you address your other concern about missed optimizations?

**Comment by gereeter** [comment]
That makes sense - that possibility was another reason that I was uncomfortable about my second commit. However, I'd be inclined to limit things to storing counts until we actually have a pass that needs the extra complexity.

My other concern about missed optimizations seemed separate from this change, so I did it in #30239 (all it took was changing `simplify_branches` to use the DFS mechanism).

**Comment by dotdash** [comment]
As the original author, I don't think I should do the review for this.

r? @nikomatsakis -- you expressed some interest in reviewing my original PR.

FWIW, my only concern here really is that re-introducing the actual predecessor map later might be harder.

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 1]
```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
...
+
+            for succ in mir.basic_block_data(bb).terminator.successors() {
+                if !seen[succ.index()] {
+                    seen[succ.index()] = true;
+                    worklist.push(*succ);
+                }
+            }
         }
 
+        // These blocks must be retained, so mark them seen even if we didn't see them
```
Comment (aravind-pg) [inline src/librustc_mir/transform/simplify_cfg.rs]: Same issue I think? These lines probably have to go before the while loop.

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 2]
```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
             Some(target)
         }
 
+        let mut predecessor_map = util::build_predecessor_map(mir);
+
         let mut changed = false;
-        for bb in mir.all_basic_blocks() {
+        let mut seen = vec![false; mir.basic_blocks.len()];
```
Comment (bluss) [inline src/librustc_mir/transform/simplify_cfg.rs]: Are the graphs large enough that a BitVector would be better for the seen map?

Comment (gereeter) [inline src/librustc_mir/transform/simplify_cfg.rs]: I don't think they are large enough for it to make a big difference. However, they might be small enough that using a `u64` or even a `u32` with no allocation might work (with a fallback to `Vec` or `BitVec`).

Regardless, I think such a change should be in a separate PR.

**Comment by nikomatsakis** [comment]
I'll take a look. For various reasons I've gotten pretty behind on reviews/e-mail/everything, but I'm working backwards from the most recent stuff. I remember that for the previous PR I was a bit unsure if this change was strictly better --- I thought there might be some cases that were optimized by the older approach that are not by the newer approach? Or do you think otherwise.

File: src/librustc_mir/transform/simplify_cfg.rs (hunk 2 above)
Comment (nikomatsakis) [inline src/librustc_mir/transform/simplify_cfg.rs]: It'd be interesting to instrument the compiler and find out what's the largest number of blocks we see in building `rustc` at least.

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 3]
```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
...
+                }
             }
 
-            mir.basic_block_data_mut(bb).terminator = terminator;
+            // See if we can merge the target block into this one
+            match terminator {
+                Terminator::Goto { target } if target.index() > DIVERGE_BLOCK.index() &&
+                        predecessor_map.num_predecessors(target) == 1 => {
+                    changed = true;
+                    let mut other_data = BasicBlockData {
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/simplify_cfg.rs]: Nit: I think this would be prettier like this

```rust
let other_data = mem::replace(mir.basic_block_data_mut(target), BasicBlockData { ... });
```

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 4]
```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
...
+                    };
+                    mem::swap(&mut other_data, mir.basic_block_data_mut(target));
+
+                    // target used to have 1 predecessor (bb), and still has only one (itself)
+                    // All the successors of target have had target replaced by bb in their
+                    // list of predecessors, keeping the number the same.
+
+                    let data = mir.basic_block_data_mut(bb);
+                    data.statements.append(&mut other_data.statements);
+                    mem::swap(&mut data.terminator, &mut other_data.terminator);
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/simplify_cfg.rs]: Why are we using swap here and not `data.terminator = other_data.terminator`?

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 5]
```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
...
+                        terminator: Terminator::Goto { target: target}
+                    };
+                    mem::swap(&mut other_data, mir.basic_block_data_mut(target));
+
+                    // target used to have 1 predecessor (bb), and still has only one (itself)
+                    // All the successors of target have had target replaced by bb in their
+                    // list of predecessors, keeping the number the same.
+
+                    let data = mir.basic_block_data_mut(bb);
+                    data.statements.append(&mut other_data.statements);
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/simplify_cfg.rs]: Nit: `data_statements.extend(other_data.statements)` seems prettier and (I think) would work fine ;) not more efficient I guess though

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 6]
```diff
@@ -65,25 +44,70 @@ impl SimplifyCfg {
...
+                    mem::swap(&mut data.terminator, &mut other_data.terminator);
+                }
+                _ => mir.basic_block_data_mut(bb).terminator = terminator
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/simplify_cfg.rs]: Nit: might be a little clearer if we move the assignment here after the match, so that it's clear that we restore the `terminator` on both paths (this took me a bit of puzzling out). But I'm not sure, and I guess it's a bit less efficient that way. Maybe just add a comment like `// restore the terminator we swapped out` on both branches?

**Comment by nikomatsakis** [comment]
Not sure what I was worried about. This seems like a simple extension. I'm inclined to r+ modulo the nits above.

**Comment by nikomatsakis** [comment]
@gereeter do you think you'll be able to address the nits I raised above, or are you too busy at the moment?

File: src/librustc_mir/transform/simplify_cfg.rs (hunk 5 above)
Comment (gereeter) [inline src/librustc_mir/transform/simplify_cfg.rs]: I'd be inclined to keep `append` because it may not be the prettiest, it is perfectly readable, and it should be more efficient.

**Comment by gereeter** [comment]
@nikomatsakis Sorry for the delay - school just finished up for the semester, so I should have more time.

It just occurred to me that this, as currently implemented, can devolve into *O(n log n)* behavior. For example, if we have

```
A -> B -> C -> D
```

then a single pass will merge `B` onto the end of `A` and `D` onto the end of `C`, but will not actually merge `AB` and `CD`. A second pass will catch this, but this takes more work than just directly merging everything into `A`. I'll post a fix shortly.

**Comment by gereeter** [comment]
Okay, nits and performance issue fixed.

**Comment by nikomatsakis** [comment]
So I think it's clear we need to devise some unit tests for these optimizations! I imagine a macro that lets you relatively easily write MIR code by hand and then observe the result.

**Comment by Aatch** [comment]
Related to the comments on #30239, it seems to me that you could just make `merge_consecutive_blocks` self-contained, working until a fixpoint. It already keeps the predecessor map up-to-date, so just make sure the unreachable blocks have already been removed before running it, then remove any newly-unreachable blocks afterwards.

It's also pointless to run `simplify_branches` after this, as it won't introduce any new branches so `simplify_branches` could probably be made self-contained too. This would make `SimplifyCfg`'s `run_on_mir` look something like this:

```rust
simplify_branches();
remove_dead_blocks();
merge_consecutive_blocks();
remove_dead_blocks();
```

**Comment by nikomatsakis** [comment]
Hmm, at first read, what @Aatch said made sense to me.

**Comment by gereeter** [comment]
@Aatch That makes lots of sense. Moreover,

- After `merge_consecutive_blocks` runs, there will be no live blocks with exactly 1 predecessor and for which that predecessor has a `Goto` terminator. Therefore, we don't even need to iterate `merge_consecutive_blocks`.
- We don't actually need to update the predecessor map in `merge_consecutive_blocks`. The only place that we do this is when shortcutting empty blocks, but `merge_consecutive_blocks` makes empty blocks dead (so they won't be merged into), and the final target at the end of the chain never could be merged into anything anyway.

**Comment by dylanmckay** [comment]
Related to #26496.

**Comment by nikomatsakis** [comment]
@gereeter so, given @Aatch's comments etc, do you plan to adjust this PR? Just trying to figure out the current status.

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with comments addressed!
